src/transform/src/reduction_pushdown.rs

```diff
@@ -81,8 +90,707 @@ impl ReductionPushdown {
+            }
+
+            if let MirRelationExpr::Join {
+                inputs,
+                equivalences,
+                demand: _,
+                implementation,
+            } = &mut **input
+            {
+                if let Some(red_pusher) = ReductionPusher::try_new(
+                    group_key,
+                    aggregates,
+                    inputs,
+                    equivalences,
+                    implementation,
+                    &[],
+                ) {
+                    if let Some(new_relation_expr) = red_pusher
+                        .get_relation_with_reduction_pushed(monotonic, expected_group_size)
+                    {
+                        *relation = new_relation_expr;
+                    }
+                }
+            } else if let MirRelationExpr::Filter { input, predicates } = &mut **input {
+                if let MirRelationExpr::Join {
+                    inputs,
+                    equivalences,
+                    demand: _,
+                    implementation,
+                } = &mut **input
+                {
+                    if let Some(red_pusher) = ReductionPusher::try_new(
+                        group_key,
+                        aggregates,
+                        inputs,
+                        equivalences,
+                        implementation,
+                        predicates,
+                    ) {
+                        if let Some(new_relation_expr) = red_pusher
+                            .get_relation_with_reduction_pushed(monotonic, expected_group_size)
+                        {
+                            *relation = new_relation_expr;
+                        }
+                    }
+                }
             }
         }
     }
 }
+
+/// Convenience struct to track the many variables involved in the reduction pushdown.
+struct ReductionPusher<'a> {
+    // Components of the new relation. Once the reduction pushdown is complete,
+    // the relation would assemble roughly like this:
+    // MirRelationExpr::Project {
+    //   input: MirRelationExpr::Map {
+    //     input: MirRelationExpr::Reduce {
+    //       input: MirRelationExpr::Join {
+    //         inputs: new_inputs,
+    //         ..
+    //         },
+    //       },
+    //       group_key: outer_group_key,
+    //       aggregates: outer_aggs,
+    //     }
+    //     scalars: downcast_partial_reduce,
+    //   }
+    //   outputs: <derived from the map inside>
+    // }
+    // except that column references in `outer_group_key` and `outer_aggs` are
+    // always in terms of the join of the old relation, and would need to be
+    // updated when assembling the new relation. In the case where
+    // an outer aggregation refers to a partial aggregate that only exist in the join of
+    // the new relation, the outer aggregation will instead refer to
+    // `MirScalarExpr::Column(old_join_arity + agg_num)`
```

Comment:
These comments sort of run in to the doctext for `new_inputs`. I wonder if we can call them out better. Will keep reading and see if it is any more than a knee-jerk reaction on my part.